tools/queuebench/main.go

```diff
@@ -0,0 +1,183 @@
+// under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	flag "github.com/spf13/pflag"
+	"go.opentelemetry.io/otel/attribute"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+
+	apmqueue "github.com/elastic/apm-queue"
+	"github.com/elastic/apm-queue/kafka"
+)
+
+const app = "queuebench"
+
+type config struct {
+	brokers    []string
+	duration   time.Duration
+	eventSize  int
+	partitions int
+	topics     []apmqueue.Topic
+	verbose    bool
+}
+
+func flags2config() *config {
+	brokers := flag.String("brokers", "", "broker/s bootstrap URL")
+	duration := flag.Int("duration", 0, "duration time in seconds for the benchmark. Required.")
+	eventSize := flag.Int("event-size", 1024, "size of each event, in bytes. Default to 1024B (1KB).")
+	partitions := flag.Int("partitions", 1, "number of partitions to use for benchmarking a single topic. Default: 1.")
+	verbose := flag.Bool("verbose", false, "print MOAR")
+
+	flag.Parse()
+
+	if *brokers == "" {
+		log.Fatal("config validation: brokers must not be empty")
+	}
+	if *duration == 0 {
+		log.Fatal("config validation: duration must be set and be greater than 0")
+	}
+
+	return &config{
+		brokers:    []string{*brokers},
+		duration:   time.Duration(*duration) * time.Second,
+		eventSize:  *eventSize,
+		partitions: *partitions,
+		topics: []apmqueue.Topic{
+			apmqueue.Topic(fmt.Sprintf("%s-%d", "run", time.Now().Unix())),
+		},
+		verbose: *verbose,
+	}
+}
+
+func main() {
+	cfg := flags2config()
+
+	log.Println("prep logger")
+	zapLevel := zap.ErrorLevel
+	if cfg.verbose {
+		zapLevel = zap.DebugLevel
+	}
+	logger, err := zap.NewDevelopment(zap.IncreaseLevel(zapLevel))
+	if err != nil {
+		log.Fatal(fmt.Sprintf("cannot create zap logger: %s", err))
+	}
+
+	log.Println("prep MeterProvider")
+	rdr := sdkmetric.NewManualReader()
+	mp := sdkmetric.NewMeterProvider(
+		sdkmetric.WithReader(rdr),
+	)
+
+	kafkaCommonCfg := kafka.CommonConfig{
+		Brokers:   cfg.brokers,
+		Namespace: app,
+
+		Logger:         logger,
+		TracerProvider: trace.NewNoopTracerProvider(),
+		MeterProvider:  mp,
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	bench := newBench(cfg)
+	bench.Setup(ctx, kafkaCommonCfg)
+	defer bench.Teardown(ctx)
+
+	log.Println("staring producer")
+	stopProducer := make(chan struct{})
+	go produce(ctx, bench.Producer, cfg, stopProducer)
+
+	log.Println("starting consumer")
+	go consume(ctx, bench.Consumer, cfg, cfg.duration*2)
+
+	var quit = make(chan struct{})
+	go func() {
+		time.Sleep(cfg.duration)
+
+		log.Println("signal producer to stop")
+		stopProducer <- struct{}{}
+		err := bench.Producer.Close()
+		log.Println("closed producer")
+		if err != nil {
+			panic(err)
+		}
+
+		cancel()
+		log.Println("gracefully stopping consumer")
+		err = bench.Consumer.Close()
+		if err != nil {
+			panic(err)
+		}
```

Comment:
https://github.com/elastic/apm-queue/issues/84 says

> Stop consumer after all events have been consumed.

How do you know that all the events have been consumed?

Reply:
Thanks to the refactoring the previous problematic behaviour is gone. I implemented two global counters to track produced and consumed ([`0931d0f` (#256)](https://github.com/elastic/apm-queue/pull/256/commits/0931d0fcaaf785e064cc6a754d5c8b30938cb5bc)) and a waiting loop that terminates on `consumed total = produced total` ([`6e8e189` (#256)](https://github.com/elastic/apm-queue/pull/256/commits/6e8e189ca8ee8bd473ae3738b7d5737f2ee0cd10))

Reply:
> At first I tried counting the number of produced messages, but then I didn't find a way to count consumed messages that I considered effective (apmqueue.Processor does not allow pointer receiver and I wasn't keen on using a global).

`apmqueue.Processor` is an interface, so you can definitely use a pointer receiver.

> I considered using otel metrics but I wasn't then sure how to account for possible consumption errors (how to stop consuming if the number of produced and consumed messages do not match for an unforeseen reason? Would a timeout be enough here?)

We could have the same issue (consumed never reaches produced) regardless of how the numbers are measured. Yes, I think having a timeout would be a good idea.

> Thanks to the refactoring the previous problematic behaviour is gone. I implemented two global counters to track produced and consumed ([0931d0f (#256)](https://github.com/elastic/apm-queue/pull/256/commits/0931d0fcaaf785e064cc6a754d5c8b30938cb5bc)) and a waiting loop that terminates on consumed total = produced total ([6e8e189 (#256)](https://github.com/elastic/apm-queue/pull/256/commits/6e8e189ca8ee8bd473ae3738b7d5737f2ee0cd10))

:+1: